src/core/reader/types/global.rs

```diff
@@ -1,28 +1,47 @@
-use crate::core::reader::types::ValType;
+use alloc::vec;
+
+use crate::core::reader::span::Span;
+use crate::core::reader::types::{ResultType, ValType};
 use crate::core::reader::{WasmReadable, WasmReader};
 use crate::execution::assert_validated::UnwrapValidatedExt;
-use crate::{unreachable_validated, Error, Result};
+use crate::globals::read_constant_instructions;
+use crate::{unreachable_validated, validate_value_stack, Error, Result};
 
-#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[derive(Debug, Copy, Clone)]
 pub struct Global {
     pub ty: GlobalType,
     // TODO validate init_expr during validation and execute during instantiation
-    pub init_expr: (),
+    pub init_expr: Span,
 }
 
 impl WasmReadable for Global {
     fn read(wasm: &mut WasmReader) -> Result<Self> {
+        let ty = GlobalType::read(wasm)?;
+        let mut init_expr = None;
+
+        let expected_type = ResultType {
+            valtypes: vec![ty.ty],
+        };
+        validate_value_stack(expected_type, |value_stack| {
+            init_expr = Some(read_constant_instructions(
+                wasm,
+                value_stack,
+                &[/* todo!(imported globals tpyes) */],
+            )?);
+
+            Ok(())
+        })?;
+
+        // At this point, we can assume that `init_expr` is `Some(_)`. `read_constant_instructions` returns a Span or an
+        // Error. If an Error is returned it is pushed up the call stack.
         Ok(Self {
-            ty: GlobalType::read(wasm)?,
-            init_expr: (), /* todo!("read constant expression") */
+            ty,
+            init_expr: init_expr.unwrap_validated(),
```

Comment:
Is there a way around unwrapping here?
Also I think this should be a normal `.unwrap()` or `.expect()` call, because for it to be `None` the `validate_value_stack` method would have to return `Ok(())` without calling the closure.

Reply:
I could create a default-value span to avoid it. The issue lies in the fact that `validate_value_stack` returns an empty result. I could technically modify it to be able to return a generic type. Or I could just replace the `unwrap_validated` with an `expect`. Your call.

Reply:
Hmm... Using a generic for the function might be misleading. Let's stick with an expect 👍